### ISSUE TITLE
one more check on climate data ... 

### ISSUE THREAD

**Issue by lizzieinvancouver** [lizzieinvancouver/ospree#311]
@cchambe12 Can you do a summer summation of GDD and compare Livneh and Princeton climate datasets for how much they agree (given same sites and years of course)? Thank you!

From Ben Cook (1 Oct 2019, showed him Ailene's figure of chilling comparisons):

> Are these correlations or R2 values (the figures say R2)? I would say this is not unreasonable, considering you are dealing with (1) daily data (which will have higher uncertainties, especially before 1950) and (2) chilling units are based on thresholds (so a few tenths of a degree of difference can be a big deal). How would the two datasets compare if, for example, you looked at April 1-September 30 GDDs?

**Comment by cchambe12** [comment]
@lizzieinvancouver Alright so I calculated GDDs from June 1st through August 31st for both sets of climate data. I made a [figure](https://github.com/lizzieinvancouver/ospree/blob/master/analyses/chilling/princetontest/livneh_vs_princeton.pdf) showing Princeton GDDs vs Livneh GDDs and a [.csv file](https://github.com/lizzieinvancouver/ospree/blob/master/analyses/chilling/princetontest/livneh_vs_princeton.csv).

**Comment by cchambe12** [comment]
@cchambe12 we're not impressed. Check out outliers and then try fitting models to see the difference in outputs, chilling diffs next. Not sure what else to do.

**Comment by lizzieinvancouver** [comment]
@cchambe12 It could be okay ... but let's get more data on the figure first if we can and see what it looks like (there should be more stations to compare? Or maybe these are al the stations we use). After that I would suggest plotting some of the raw data maybe ... or we can contact Ben again for suggestions (but first we need more points if easy). Thanks for working on this!

**Comment by cchambe12** [comment]
@lizzieinvancouver In the first set of data I just used the stations and years present in the OSPREE data. I only used datasets that were before 2014 since Livneh doesn't extend past this. [Here](https://github.com/lizzieinvancouver/ospree/blob/master/analyses/chilling/princetontest/compare_gdd_tmin_tmax.pdf) is a new figure where I generated a list of 1000 latxlongs for different made up years. Calculated GDDs are in the left panel, Tmin in the middle, and Tmax to the right. It looks like the reason for the almost linear trend of high princeton GDDs may be due to Tmax values. It looks like the princeton Tmax values are cut off at 10C... 

Additional considerations: livneh is 0.1 x 0.1 whereas Princeton is 0.25 x 0.25 which may be the reason for the discrepancies, especially if in regions that have more drastic elevation differences.

**Comment by cchambe12** [comment]
@cchambe12 manually check 5 sites to check code

**Comment by cchambe12** [comment]
Okay, I did a deeper look into the climate data and checked on 5 random sites from my dataset to make sure the code looks okay. 

I added in leap years into the code just in case but since we are just looking at summer days, this had no effect. 

Major takeaways:
1. Livneh has a lot more missing data
2. Princeton is generally higher in Tmin and Tmax
3. Livneh generally has a bigger difference between Tmin and Tmax, whereas it can be very minimal for some sites on Princeton

Otherwise, I think the code looks okay... unless I'm converting Kelvin to Celsius wrong! Kelvin - 273.15 = Celsius?

**Comment by lizzieinvancouver** [comment]
From 18 Nov 2019 email from Park Williams ...

> Hi Lizzie,
> 
> Yikes, this comparison ins’t very promising at all is it? Are these scatter plots of daily data for all North American grid cells, or just a region? And for 1948-2015 period, or did you use the older Princeton product that goes back to 1901? And did you upscale the 1/8-deggree Livneh data to the 1/4-degree resolution of the Princeton data?
> 
> Overall I wouldn’t have been surprised to see that they don’t agree that well, but these scatter plots look like there’s essentially no agreement. It would be helpful to see maps of correlation to see where they agree better and where they agree worse. And it would also be helpful to break these into 1915-1947, 1948-1978, and 1979-2015.
> 
> The Princeton dataset is a bit of a black box, but I believe is largely reanalysis based. I think Livneh’s dataset is probably better for a study specific to North America, as it’s based on stations and the energy to make this dataset was focussed on North America. But if you’re using wind speed and considering years prior to 1948, note that Livneh assigns the climatological mean annual cycle to wind speed prior to 1948.
> 
> Park
> 
> 
> -----------------------------------------------------------------------
> Park Williams
> Lamont Associate Research Professor, Junior Staff
> Lamont-Doherty Earth Observatory of Columbia University
> 61 Rt 9W; Palisades, NY 10964
> williams@ldeo.columbia.edu

**Comment by cchambe12** [comment]
@cchambe12 find problem areas and plot rasters

**Comment by cchambe12** [comment]
More updates on the climate saga:

@cchambe12 Try to tackle some of Park's additional suggestions for a more thorough analysis

Breakdown pre vs post 1979 and remove monthly climatological means (de-seasonalized data)

> Thank goodness! If you’re interested in digging any deeper, I’d still be interested in seeing this broken out by pre- vs post-1979, and also de-seasonalized (monthly climatological means removed).

Park


-----------------------------------------------------------------------
Park Williams
Lamont Associate Research Professor, Junior Staff
Lamont-Doherty Earth Observatory of Columbia University
61 Rt 9W; Palisades, NY 10964
williams@ldeo.columbia.edu

On Nov 21, 2019, 1:06 PM -0500, Chamberlain, Catherine <cchamberlain@g.harvard.edu>, wrote:
Hi Park and Ben,

Thank you for all of your feedback and suggestions. I mapped out the raster files I discovered that the longitudes were from 0 to 360 in the princeton data when I was treated them as -180 to 180. After adjusting for this, I reran the comparison and I appears we actually have agreement (see attached figure).

Sorry if this caused concern!

All the best,
Cat

On Wed, Nov 20, 2019 at 9:58 AM Elizabeth M Wolkovich <wolkovic@mail.ubc.ca> wrote:
Hi Park and Ben,

Thanks! We're using the newer Pinceton data and the draws are 1000
random ones across North America. We didn't upscale, we just picked
points and took the nearest cell for Livneh and for Princeton (should
introduce some error, but feels like not this much). We just need
temperature data (daily min and max), but we do need temperatures
through at least 2016 so we really can't use Livneh.

We're doing some visualizations to see if we can figure anything.

If either of you have recommendations of easy ways to load and look at
the data, please let me know. Ben, I feel like I have often seen you use
some program that loads and quickly visualizes these sorts of data so if
you can let me know what it is we might do that as well -- as a third
check on our code.

Thanks again.

All the best,
Lizzie

**Comment by cchambe12** [comment]
@MoralesCastilla @lizzieinvancouver @dbuona @AileneKane @legault @FaithJones @DeirdreLoughnan @DarwinSodhi @md-garner 

The climate data seems to be done... But we are losing more rows of data than we used to after running the bb_analysis/cleaning/clean_bbperctodays.R. I think it may have to do with the photoperiod column? Have we always had clean_photoperiod.R after clean_bbperctodays.R? Otherwise, I really can't figure out why we are losing data that we used to have. One study I am working with to compare bbculdesac to master is falusi96. Any help on this would be much appreciated!!

**Comment by cchambe12** [comment]
@lizzieinvancouver @dbuona @MoralesCastilla @AileneKane @FaithJones @legault @DeirdreLoughnan @DarwinSodhi @md-garner 

Looking for a friend! I think I have finally found the issue... bb_analysis/cleaning/clean_ambientforcingfromdailyclimate.R lines 129-139. The calculated forcing temperatures are not being appended to the main dataframe at all. I haven't used this approach before so I'm struggling to see through where the issue is. Let me know if you're interested in helping out and I will stop looking at it to avoid merge conflicts. Thanks!!

**Comment by lizzieinvancouver** [comment]
@AileneKane Would you be able to look at this sometime? Cat and I worked on this today and the indexing in bb_analysis/cleaning/clean_ambientforcingfromdailyclimate.R lines 129-139 does seem to be returning all NAs, but I could not tell if it was because some rows were deleted so the indexing no longer works or something else. Any help would be much appreciated!

**Comment by AileneKane** [comment]
@cchambe12 @lizzieinvancouver yes! I will look at this in the next few days!

**Comment by AileneKane** [comment]
@lizzieinvancouver @cchambe12 a question before I start looking at this: is the problem only with  livenh, only princeton, or both?

**Comment by lizzieinvancouver** [comment]
@AileneKane My understanding is that it worked in your code (i.e., with Livneh), but is not working with Princeton.

**Comment by cchambe12** [comment]
@AileneKane That's correct! Worked with bbculdesac but is no longer working for Princeton. It seems to be that the new data frame I am building with Princeton has different row names

**Comment by AileneKane** [comment]
@lizzieinvancouver @cchambe12 When I ran this code, it seemed that lines 129-139 returned NAs for about half of bb uniqueIDs (558/1015). This was because there were no climate data for these unique IDs.  Is this what you also found?

**Comment by cchambe12** [comment]
@AileneKane For some reason I am only getting avg_bbtemp for 73 out of 11,959 rows of data... I've tried two computers now and am getting the same thing but it sounds like it is working for you - at least working better for you.

**Comment by cchambe12** [comment]
@AileneKane @lizzieinvancouver 

So, if I try to merge avg_bbtemp to d without indexing I can run the following lines:

`## saving results to output - d
bb.sub <- subset(bb, select=-c(expstartdate, response.time.integer, bbdate, lastchilldate, uniqueID))
d <- full_join(d, bb.sub)`

Which works! The dimensions are the same and we retain all of the unique values for bb$avg_bbtemp (1560). I know we like to try and avoid dplyr functions so I'd be happy to come up with an alternative if we don't like this solution or see possible issues arising.

**Comment by AileneKane** [comment]
@cchambe12 @lizzieinvancouver No, it's not working for me. I am only getting avg_bbtemp for 72 out of 11,959 rows of data in d. I was just curious if you are concerned about the many sites for which we do not have climate data, or if that makes sense to you! I'm looking at this more about the indexing.

**Comment by AileneKane** [comment]
@cchambe12 @lizzieinvancouver It is clear that the indexing is not working. It selects the same number of rows from both bb and d, but the rows it selects are not equivalent because the two dataframes are not in the same order (e.g., row 2 in bb is in d, but it is not row 2 in d- it is row 5). I'm actually surprised that it worked in bbculdesac- we should check this. 
The choices to solve this that I can think of are:
1) Sort the two dataframes so that they are in the same order prior to indexing in the folloing lines
indexbb <- which(uniquevalsbb %in% uniquevalsd)
indexd <- which(uniquevalsd %in% uniquevalsbb)
2) Use merge or join instead of indexes.
Let me know if you want help coding the solution! I don't have a preference for merge vs join so I leave that to you and Lizzie to decide if you don't want to use dplyr.

**Comment by cchambe12** [comment]
Thanks @AileneKane! I will triple check bbculdesac but when I was trying to solve this problem on master it seemed like it was working just fine. I'll make sure the lines were matching up appropriately though. I was unable to make option 1 (with the indexing) work so I went with join in dplyr. @lizzieinvancouver is this okay or would you rather I used merge?

**Comment by cchambe12** [comment]
@lizzieinvancouver I can't figure out merge!! Not sure why I always have so much trouble with it. Check analyses/bb_analyses/cleaning/clean_ambientforcingfromdailyclimate.R lines127-132. I have notes. I can get `goo' to work but that's using full_join otherwise merge subsets the dataframe to only rows that have ambient forcing calculated.

**Comment by lizzieinvancouver** [comment]
@lizzieinvancouver Should do this! @cchambe12 Please pester me about it in February.
